src/test/java/io/jenkins/plugins/httpclient/RobustHTTPClientTest.java

```diff
@@ -24,19 +24,110 @@
+import static org.junit.Assert.assertThrows;
 
+import hudson.AbortException;
+import hudson.model.TaskListener;
+import java.io.File;
+import java.io.IOException;
 import java.net.URL;
+import java.nio.file.Files;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.junit.ClassRule;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.WithoutJenkins;
 
 public class RobustHTTPClientTest {
 
+    @ClassRule
+    public static JenkinsRule j = new JenkinsRule();
+
+    @Rule
+    public TemporaryFolder tempFolder = new TemporaryFolder();
+
     @Test
+    @WithoutJenkins
     public void sanitize() throws Exception {
-        assertEquals("http://x.com/some/path", RobustHTTPClient.sanitize(new URL("http://x.com/some/path")));
-        assertEquals(
-                "https://x.com/some/path?…", RobustHTTPClient.sanitize(new URL("https://x.com/some/path?auth=s3cr3t")));
-        assertEquals(
-                "https://…@x.com/otherpath", RobustHTTPClient.sanitize(new URL("https://user:s3cr3t@x.com/otherpath")));
+        assertThat(
+                RobustHTTPClient.sanitize(new URL("http://example.com/some/long/path")),
+                is("http://example.com/some/long/path"));
+        assertThat(
+                RobustHTTPClient.sanitize(new URL("https://example.com/some/path?auth=s3cr3t")),
+                is("https://example.com/some/path?…"));
+        assertThat(
+                RobustHTTPClient.sanitize(new URL("https://user:s3cr3t@example.com/otherpath")),
+                is("https://…@example.com/otherpath"));
+    }
+
+    @Test
+    @WithoutJenkins
+    public void sanitizeThrowsException() {
+        final AssertionError e = assertThrows(AssertionError.class, () -> {
+            RobustHTTPClient.sanitize(new URL("https://example.com/ /has/space/in/url/"));
+        });
+        assertThat(e.getCause().getMessage(), containsString("Illegal character in path at index 20"));
+    }
+
+    @Test
+    public void testDownloadFile() throws Exception {
+        JenkinsRule.WebClient wc = j.createWebClient();
+        RobustHTTPClient client = wc.executeOnServer(() -> {
+            return new RobustHTTPClient();
+        });
+        File f = new File(tempFolder.newFolder(), "jenkins-index.html");
+        client.downloadFile(f, j.getURL(), TaskListener.NULL);
+        assertThat(Files.readString(f.toPath()), containsString("<title>Dashboard [Jenkins]</title>"));
+    }
+
+    @Test
+    public void testDownloadFileWithTimeoutException() throws Exception {
+        JenkinsRule.WebClient wc = j.createWebClient();
+        RobustHTTPClient client = wc.executeOnServer(() -> {
+            RobustHTTPClient internalClient = new RobustHTTPClient();
+            internalClient.setStopAfterAttemptNumber(1);
+            internalClient.setTimeout(50, TimeUnit.MICROSECONDS);
```

Comment:
This looks flaky. Set a real timeout, and instead of downloading the Jenkins root URL, use a `@TestExtension` on an `UnprotectedRootAction` whose `doIndex` sleeps.

Reply:
Thanks.  I like that idea much better.

Reply:
I won't have time to implement that change immediately, since it will require that I learn more about `@TestExtension` and `UnprotectedRootAction`.  I'll leave this pull request open until I have time to implement that.

Reply:
Roughly, and from memory:

```java
@Test public void xxx() {
  theClient.download(r.getRootURL() + "hangs/");
}
@TestExtension("xxx") public static final class Hangs extends InvisibleAction implements UnprotectedRootAction {
  @Override public String getUrl() {return "hangs";}
  HttpResponse public void doIndex() {Thread.sleep(Long.MAX_VALUE);}
}
```

Reply:
Implemented in c21f9c1ca109ea04a892dab7b065588abf675e34